File: evals/models/compare.ts
import { appendFileSync, existsSync, readFileSync } from 'node:fs'
import type { Chunk, EmbeddingsIndex } from '../lib/corpus.ts'
import { readJsonLines } from '../lib/jsonl.ts'
import { unpackVectors } from '../lib/vectors.ts'
import { denseRetriever } from '../retrievers/dense.ts'
import { assertDenseInputsFresh, type Labelled } from '../run-retrieval.ts'

/**
 * Runs two models over the same questions and the same retrieved context, and records what each
 * one spent. The design is fixed in evals/results/thinkingcap-preregistration.md; this file only
 * collects data, and deliberately computes nothing — no ratios, no verdicts. Analysis happens
 * after judging, from the recorded rows.
 *
 * The hard set is the pre-registered sample because those queries are diagnostic rather than
 * lookup: a task with no reasoning in it cannot test a claim about reasoning tokens. The bulk set
 * runs as a separate, larger sample — its queries are lexically closer to their targets, so fewer
 * items should collapse into ties.
 */

/**
 * Which label set to run. The hard set was the pre-registered sample; the bulk set is the planned
 * extension, kept in its own file because the two are never averaged — bulk queries are generated
 * paraphrases and hard queries are written from memory, which are different tasks.
 */
const SET = process.env.EVAL_SET === 'bulk' ? 'bulk' : 'hard'
/**
 * Names the serving conditions, not just the sample. The first runs went through an aggregator that
 * routed across hosts of unknown quantisation; these go to two endpoints we control. Mixing the two
 * in one file would be undetectable afterwards, because nothing in a row says which.
 */
const RUN_TAG = process.env.EVAL_RUN_TAG ?? 'hfe'
const OUT = `evals/data/model-runs.${RUN_TAG}.${SET}.jsonl`
/**
 * The sampling the model's authors used for the numbers under replication. Greedy decoding was the
 * original design here; it measures the model outside its documented regime and risks the
 * repetition loops reasoning models fall into — which is what a ceiling-length completion in the
 * first runs looks like. See amendment 6.
 */
const TEMPERATURE = Number.parseFloat(process.env.MODEL_TEMPERATURE ?? '1.0')
const TOP_P = Number.parseFloat(process.env.MODEL_TOP_P ?? '0.95')
const SAMPLE_TOP_K = Number.parseInt(process.env.MODEL_TOP_K ?? '20', 10)
const MIN_P = Number.parseFloat(process.env.MODEL_MIN_P ?? '0')
/**
 * Sampling makes one run a draw rather than a measurement — the authors report high run-to-run
 * variability and used five seeds. Each seed is a full pass over the set and the spread across them
 * is what gets reported, so a single-seed number would be an anecdote with a confidence interval.
 */
const SEEDS = (process.env.MODEL_SEEDS ?? '1,2,3').split(',').map((x) => Number.parseInt(x.trim(), 10))
/**
 * Restricts a run to one endpoint. Alternating between them per question needs both machines up
 * for the whole run, and the slower model then bills the faster one's idle time — on dedicated
 * hardware that doubles the cost of the same calls. The prompts do not depend on which endpoint
 * ran, so splitting the passes changes the bill and nothing else.
 */
const ONLY = process.env.EVAL_ONLY
/**
 * Fixed at bge-m3 top-5 even after the retriever improves. Changing retrieval between the two
 * halves would confound a better context with a larger sample; retrieval quality is measured by
 * the retrieval eval, not by this one.
 */
const TOP_K = 5
/**
 * Deliberately generous. A low ceiling truncates whichever model thinks longer — which is the base
 * model by hypothesis — so the saving under test would partly be this setting rather than the
 * fine-tune. Raised to the larger of the two budgets the authors used, because a truncated item is
 * dropped and the dropped ones are the hardest questions, not a random sample of them.
 */
const MAX_TOKENS = Number.parseInt(process.env.MODEL_MAX_TOKENS ?? '24576', 10)

export type Run = {
  query: string
  model: string
  /**
   * Which pass this row came from. Sampling makes one run a draw, so the same question is asked
   * once per seed and the spread across them is the measurement.
   */
  seed: number
  answer: string
  reasoningTokens: number | null
  /**
   * Length of the reasoning trace in characters. Recorded because the token count cannot be
   * trusted across these two providers: Sference returns the trace in `reasoning_content` while
   * reporting `reasoning_tokens: 0`. Characters need no tokenizer and both traces are Czech.
   */
  reasoningChars: number
  completionTokens: number | null
  promptTokens: number | null
  /** Recorded for completeness, not reported: it measures the provider's serving stack. */
  latencyMs: number
  finishReason: string | null
  /**
   * Who actually served the call. An aggregator routes freely across hosts differing in
   * quantisation and hardware; the first runs recorded neither field, so that is now unknowable.
   */
  provider: string | null
  generationId: string | null
  contextIds: string[]
}

const SYSTEM = [
  'Odpovídej česky a pouze na základě přiložených úryvků.',
  'Když v nich odpověď není, napiš to — nedomýšlej si.',
].join(' ')

export function buildPrompt(query: string, context: Chunk[]): string {
  const blocks = context.map((c, i) => `--- úryvek ${i + 1} (${c.ts.slice(0, 10)}) ---\n${c.text}`)
  return `${blocks.join('\n\n')}\n\nOTÁZKA: ${query}`
}

/**
 * A run is identified by the pair, so an interrupted session resumes without paying twice. Both
 * sides of the resume check must build the key the same way — they did not, and the mismatch cost
 * a duplicate call per already-answered question without anything failing.
 *
 * The separator is NUL because a model name or a query may contain a space, and it is written as an
 * escape because a literal one makes git treat this whole file as binary.
 */
export function runKey(model: string, query: string, seed = 0): string {
  return [seed, model, query].join('\u0000')
}

export function doneKeys(path: string): Set<string> {
  return new Set(readJsonLines<Run>(path).map((r) => runKey(r.model, r.query, r.seed ?? 0)))
}

type Usage = {
  completion_tokens?: number
  prompt_tokens?: number
  completion_tokens_details?: { reasoning_tokens?: number }
  // Some providers report reasoning separately at the top level instead.
  reasoning_tokens?: number
}

/**
 * Reasoning tokens are the whole point and the field is not standardised, so both known shapes are
 * read and a miss returns null rather than zero. Zero would silently become "no reasoning at all"
 * in the median and make the saving look total.
 */
export function reasoningTokensOf(usage: Usage | undefined): number | null {
  const v = usage?.completion_tokens_details?.reasoning_tokens ?? usage?.reasoning_tokens
  return typeof v === 'number' ? v : null
}

/**
 * The trace itself, under either name the two providers use. OpenRouter puts it in `reasoning`,
 * Requesty in `reasoning_content` — and the latter pairs it with `reasoning_tokens: 0`, so a run
 * trusting the count alone would record a model that thought for 143 tokens as having not thought
 * at all, and report a 100% saving.
 */
export function reasoningTextOf(message: { reasoning?: string; reasoning_content?: string } | undefined): string {
  return message?.reasoning ?? message?.reasoning_content ?? ''
}

const THINK_OPEN = '<think>'
const THINK_CLOSE = '</think>'

/**
 * Pulls the trace out of the content itself. A plain vLLM server started without a reasoning
 * parser returns the whole `<think>…</think>` block inline, so a run that only read the provider
 * fields would score the trace as part of the answer and measure a saving of zero.
 *
 * Doing the split here rather than asking the server to do it also removes the dependency that
 * already cost this experiment: the same code now derives the trace for every endpoint, instead of
 * trusting fields that two providers filled in differently and one filled in wrongly.
 */
export function splitThinking(content: string): { answer: string; thinking: string } {
  const thinking: string[] = []
  let answer = ''
  // This model's chat template puts the opening tag in the generation prompt, so the completion
  // begins inside the block and carries only the closing one. Supplying the missing tag rather
  // than special-casing it keeps one code path for both shapes.
  const firstClose = content.indexOf(THINK_CLOSE)
  const firstOpen = content.indexOf(THINK_OPEN)
  let rest = firstClose !== -1 && (firstOpen === -1 || firstClose < firstOpen) ? THINK_OPEN + content : content
  while (true) {
    const open = rest.indexOf(THINK_OPEN)
    if (open === -1) {
      answer += rest
      break
    }
    answer += rest.slice(0, open)
    const after = rest.slice(open + THINK_OPEN.length)
    const close = after.indexOf(THINK_CLOSE)
    // No closing tag means the ceiling cut the model off mid-thought. The fragment is reasoning,
    // not an answer, and letting it through would enter a stream of thought into the data as one.
    if (close === -1) {
      thinking.push(after.trim())
      break
    }
    thinking.push(after.slice(0, close).trim())
    rest = after.slice(close + THINK_CLOSE.length)
  }
  return { answer: answer.trim(), thinking: thinking.join('\n') }
}

/**
 * Decides what the model answered and what it merely thought, across every shape these endpoints
 * produce. Kept separate from the HTTP call so it can be tested: every measurement error this
 * experiment has suffered so far lived in exactly this step and failed silently.
 */
export function separateTrace(
  message: { content?: string; reasoning?: string; reasoning_content?: string } | undefined,
  finishReason: string | null,
): { answer: string; reasoningText: string } {
  const content = message?.content ?? ''
  const fromField = reasoningTextOf(message)
  // A completion the ceiling cut off mid-thought never reaches its closing tag, and because the
  // opening one lives in the prompt it arrives with no tags at all — shaped exactly like a plain
  // answer. Recorded as one it would report a long answer and no reasoning: the saving inverted.
  if (finishReason === 'length' && !fromField && !content.includes(THINK_CLOSE)) {
    return { answer: '', reasoningText: content.trim() }
  }
  const split = splitThinking(content)
  return { answer: split.answer, reasoningText: fromField || split.thinking }
}

type Reply = {
  answer: string
  reasoningText: string
  usage: Usage | undefined
  finishReason: string | null
  latencyMs: number
  provider: string | null
  generationId: string | null
}

/**
 * Each model carries its own endpoint. The base model and the fine-tune are not served by the same
 * provider, which is a confound the pre-registration records rather than hides: quantisation and
 * sampling implementation differ per host, so a quality difference is not purely the model's.
 * Token counts survive this better than quality does — the tokenizer is shared.
 */
export type Endpoint = { label: 'base' | 'tuned'; base: string; key: string; model: string }

export function endpointsFromEnv(env: NodeJS.ProcessEnv): Endpoint[] {
  const read = (p: string, label: 'base' | 'tuned'): Endpoint => {
    const base = env[`${p}_API_BASE`]
    // Both endpoints are now our own, authenticated by the one account token, so the per-endpoint
    // key is optional rather than a third copy of the same secret in the same file.
    const key = env[`${p}_API_KEY`] ?? env.HF_TOKEN
    const model = env[`${p}_MODEL`]
    if (!base || !key || !model) {
      throw new Error(`set ${p}_API_BASE, ${p}_API_KEY and ${p}_MODEL`)
    }
    return { label, base, key, model }
  }
  return [read('BASE', 'base'), read('TUNED', 'tuned')].filter((e) => !ONLY || e.label === ONLY)
}

/**
 * Transport-level retry, and only that. A run is ninety calls of up to two minutes each, so a
 * dropped connection is a matter of when rather than whether — one ECONNRESET already threw away
 * an hour of a paid run at call twenty-five. Retried on network failures and on the statuses that
 * mean "ask again"; a 4xx is a request that will fail identically next time and stops the run.
 *
 * Deliberately does not cover an empty completion. That is a reply the server meant to send, and
 * retrying it would quietly resample until the model said something — which is a different
 * experiment from the one being run.
 */
async function fetchWithRetry(url: string, init: RequestInit, label: string): Promise<Response> {
  const waits = [2000, 8000, 30000]
  for (let attempt = 0; ; attempt++) {
    try {
      const res = await fetch(url, init)
      if (res.status !== 429 && res.status < 500) {
        return res
      }
      if (attempt >= waits.length) {
        return res
      }
      console.log(`  ${label}: HTTP ${res.status}, zkouším znovu za ${waits[attempt] / 1000}s`)
    } catch (err) {
      if (attempt >= waits.length) {
        throw err
      }
      console.log(`  ${label}: ${(err as Error).message}, zkouším znovu za ${waits[attempt] / 1000}s`)
    }
    await new Promise((r) => setTimeout(r, waits[attempt]))
  }
}

async function ask(prompt: string, ep: Endpoint, seed: number): Promise<Reply> {
  const t0 = performance.now()
  const res = await fetchWithRetry(
    `${ep.base.replace(/\/$/, '')}/chat/completions`,
    {
      method: 'POST',
      headers: { authorization: `Bearer ${ep.key}`, 'content-type': 'application/json' },
      body: JSON.stringify({
        model: ep.model,
        // Asks the provider to report the reasoning trace and its token count. Affects what comes
        // back, not what is generated, so both sides can carry it without becoming different runs.
        include_reasoning: true,
        temperature: TEMPERATURE,
        top_p: TOP_P,
        top_k: SAMPLE_TOP_K,
        min_p: MIN_P,
        seed,
        max_tokens: MAX_TOKENS,
        messages: [
          { role: 'system', content: SYSTEM },
          { role: 'user', content: prompt },
        ],
      }),
    },
    ep.model,
  )
  if (!res.ok) {
    throw new Error(`${ep.model} ${res.status}: ${(await res.text()).slice(0, 400)}`)
  }
  const data = (await res.json()) as {
    id?: string
    provider?: string
    choices?: Array<{
      message?: { content?: string; reasoning?: string; reasoning_content?: string }
      finish_reason?: string
    }>
    usage?: Usage
  }
  const choice = data.choices?.[0]
  const finishReason = choice?.finish_reason ?? null
  const { answer, reasoningText } = separateTrace(choice?.message, finishReason)
  // Hitting the ceiling is a foreseen outcome the pre-registration handles by excluding the item,
  // so it is recorded and the run continues. An empty completion for any other reason is a broken
  // reply and stops the run rather than entering the data as a valid short answer.
  if (!answer.trim() && finishReason !== 'length') {
    throw new Error(`${ep.model} returned no content (finish_reason ${finishReason})`)
  }
  return {
    answer,
    reasoningText,
    usage: data.usage,
    finishReason,
    latencyMs: performance.now() - t0,
    provider: data.provider ?? null,
    generationId: data.id ?? null,
  }
}

async function main(): Promise<void> {
  const endpoints = endpointsFromEnv(process.env)

  const chunks = readJsonLines<Chunk>('evals/data/corpus.jsonl', { required: true })
  const labels = readJsonLines<Labelled>('evals/data/retrieval.labels.jsonl', { required: true })
  const selected = labels.filter((l) => l.kind === SET)

  const meta = JSON.parse(readFileSync('evals/data/embeddings.index.json', 'utf8')) as EmbeddingsIndex
  const vectors = unpackVectors(readFileSync('evals/data/embeddings.bin'), meta.dim, meta.ids.length)
  const qv = JSON.parse(readFileSync('evals/data/query-vectors.json', 'utf8')) as Record<string, number[]>
  assertDenseInputsFresh(
    chunks,
    meta,
    labels.map((l) => l.query),
    qv,
  )
  const dense = denseRetriever(meta.ids, vectors, (q) => new Float32Array(qv[q]))
  const byId = new Map(chunks.map((c) => [c.id, c]))

  const done = existsSync(OUT) ? doneKeys(OUT) : new Set<string>()
  let ran = 0
  let missingReasoning = 0
  let truncated = 0

  // Seed outermost, so an interrupted run leaves whole comparable passes behind rather than a set
  // covered unevenly across seeds.
  const passes = SEEDS.flatMap((seed) => selected.map((label) => ({ seed, label })))
  for (const { seed, label } of passes) {
    const contextIds = dense.search(label.query, TOP_K)
    const prompt = buildPrompt(
      label.query,
      contextIds.map((id) => byId.get(id)!),
    )
    for (const ep of endpoints) {
      if (done.has(runKey(ep.model, label.query, seed))) {
        continue
      }
      const { answer, reasoningText, usage, finishReason, latencyMs, provider, generationId } = await ask(
        prompt,
        ep,
        seed,
      )
      const reasoningTokens = reasoningTokensOf(usage)
      // Zero tokens alongside a non-empty trace is the provider under-reporting, not a model that
      // skipped thinking — counted separately so it cannot pass as a saving.
      if (reasoningTokens === null || (reasoningTokens === 0 && reasoningText.length > 0)) {
        missingReasoning++
      }
      if (finishReason === 'length') {
        truncated++
      }
      const row: Run = {
        query: label.query,
        model: ep.model,
        seed,
        answer,
        reasoningTokens,
        reasoningChars: reasoningText.length,
        completionTokens: usage?.completion_tokens ?? null,
        promptTokens: usage?.prompt_tokens ?? null,
        latencyMs,
        finishReason,
        provider,
        generationId,
        contextIds,
      }
      appendFileSync(OUT, `${JSON.stringify(row)}\n`)
      ran++
      console.log(
        `${ran} [seed ${seed}] ${ep.model} | reasoning ${reasoningTokens ?? '—'} tok / ${reasoningText.length} zn` +
          ` | completion ${row.completionTokens ?? '—'}${finishReason === 'length' ? ' | USEKNUTO' : ''}`,
      )
    }
  }

  console.log(`\nhotovo: ${SET} set, seedy ${SEEDS.join(',')}, ${ran} běhů zapsáno do ${OUT}`)
  if (missingReasoning > 0) {
    console.log(`POZOR: ${missingReasoning} odpovědí s nedůvěryhodným reasoning_tokens.`)
    console.log('Použij reasoningChars; viz thinkingcap-preregistration.md, dodatek 3.')
  }
  if (truncated > 0) {
    console.log(`POZOR: ${truncated} odpovědí useknuto stropem ${MAX_TOKENS} tokenů.`)
    console.log('Ty otázky se ze srovnání vyřazují — viz dodatek 2, ne že by šetřily tokeny.')
  }
}

if (/[\\/]models[\\/]compare\.ts$/.test(process.argv[1] ?? '')) {
  await main()
}

File: evals/models/compare.py
import json
import os
import time
from collections.abc import Mapping
from pathlib import Path
from typing import Literal, NamedTuple, TypedDict

from lib.corpus import CORPUS, Chunk, EmbeddingsIndex
from lib.env import load_dotenv
from lib.http import HttpError, dig, post_json_with_retry
from lib.jsonl import append_json_line, read_json_lines
from lib.paths import DATA
from lib.vectors import unpack_vectors
from retrievers.dense import DenseRetriever
from run_retrieval import Labelled, assert_dense_inputs_fresh

SET = 'bulk' if os.environ.get('EVAL_SET') == 'bulk' else 'hard'
RUN_TAG = os.environ.get('EVAL_RUN_TAG', 'hfe')
OUT = DATA / f'model-runs.{RUN_TAG}.{SET}.jsonl'
TEMPERATURE = float(os.environ.get('MODEL_TEMPERATURE', '1.0'))
TOP_P = float(os.environ.get('MODEL_TOP_P', '0.95'))
SAMPLE_TOP_K = int(os.environ.get('MODEL_TOP_K', '20'))
MIN_P = float(os.environ.get('MODEL_MIN_P', '0'))
SEEDS = [int(x) for x in os.environ.get('MODEL_SEEDS', '1,2,3').split(',')]
ONLY = os.environ.get('EVAL_ONLY')
TOP_K = 5
MAX_TOKENS = int(os.environ.get('MODEL_MAX_TOKENS', '24576'))

SYSTEM = ' '.join(
    [
        'Odpovídej česky a pouze na základě přiložených úryvků.',
        'Když v nich odpověď není, napiš to — nedomýšlej si.',
    ]
)

THINK_OPEN = '<think>'
THINK_CLOSE = '</think>'


class Run(TypedDict):
    query: str
    model: str
    seed: int
    answer: str
    reasoningTokens: int | None
    reasoningChars: int
    completionTokens: int | None
    promptTokens: int | None
    latencyMs: float
    finishReason: str | None
    provider: str | None
    generationId: str | None
    contextIds: list[str]


def build_prompt(query: str, context: list[Chunk]) -> str:
    blocks = [f'--- úryvek {i} ({c.ts[:10]}) ---\n{c.text}' for i, c in enumerate(context, start=1)]
    return '\n\n'.join(blocks) + f'\n\nOTÁZKA: {query}'


def run_key(model: str, query: str, seed: int = 0) -> str:
    return '\0'.join([str(seed), model, query])


def done_keys(path: Path) -> set[str]:
    return {run_key(r['model'], r['query'], r.get('seed', 0)) for r in read_json_lines(path)}


def reasoning_tokens_of(usage: Mapping[str, object] | None) -> int | None:
    v = dig(usage, 'completion_tokens_details', 'reasoning_tokens')
    if v is None:
        v = dig(usage, 'reasoning_tokens')
    return v if isinstance(v, int) and not isinstance(v, bool) else None


def reasoning_text_of(message: Mapping[str, object] | None) -> str:
    text = dig(message, 'reasoning')
    if not isinstance(text, str):
        text = dig(message, 'reasoning_content')
    return text if isinstance(text, str) else ''


class Split(NamedTuple):
    answer: str
    thinking: str


def split_thinking(content: str) -> Split:
    thinking: list[str] = []
    answer = ''
    first_close = content.find(THINK_CLOSE)
    first_open = content.find(THINK_OPEN)
    rest = (
        THINK_OPEN + content
        if first_close != -1 and (first_open == -1 or first_close < first_open)
        else content
    )
    while True:
        open_at = rest.find(THINK_OPEN)
        if open_at == -1:
            answer += rest
            break
        answer += rest[:open_at]
        after = rest[open_at + len(THINK_OPEN) :]
        close_at = after.find(THINK_CLOSE)
        if close_at == -1:
            thinking.append(after.strip())
            break
        thinking.append(after[:close_at].strip())
        rest = after[close_at + len(THINK_CLOSE) :]
    return Split(answer.strip(), '\n'.join(thinking))


class Separated(NamedTuple):
    answer: str
    reasoning_text: str


def separate_trace(message: Mapping[str, object] | None, finish_reason: str | None) -> Separated:
    content = dig(message, 'content')
    content = content if isinstance(content, str) else ''
    from_field = reasoning_text_of(message)
    if finish_reason == 'length' and not from_field and THINK_CLOSE not in content:
        return Separated('', content.strip())
    split = split_thinking(content)
    return Separated(split.answer, from_field or split.thinking)


Label = Literal['base', 'tuned']


class Endpoint(NamedTuple):
    label: Label
    base: str
    key: str
    model: str


def endpoints_from_env(env: Mapping[str, str]) -> list[Endpoint]:
    def read(prefix: str, label: Label) -> Endpoint:
        base = env.get(f'{prefix}_API_BASE')
        key = env.get(f'{prefix}_API_KEY') or env.get('HF_TOKEN')
        model = env.get(f'{prefix}_MODEL')
        if not base or not key or not model:
            raise SystemExit(f'set {prefix}_API_BASE, {prefix}_API_KEY and {prefix}_MODEL')
        return Endpoint(label, base, key, model)

    only = env.get('EVAL_ONLY')
    return [e for e in (read('BASE', 'base'), read('TUNED', 'tuned')) if not only or e.label == only]


class Reply(NamedTuple):
    answer: str
    reasoning_text: str
    usage: Mapping[str, object] | None
    finish_reason: str | None
    latency_ms: float
    provider: str | None
    generation_id: str | None


def ask(prompt: str, ep: Endpoint, seed: int) -> Reply:
    t0 = time.perf_counter()
    try:
        data = post_json_with_retry(
            f'{ep.base.rstrip("/")}/chat/completions',
            {
                'model': ep.model,
                'include_reasoning': True,
                'temperature': TEMPERATURE,
                'top_p': TOP_P,
                'top_k': SAMPLE_TOP_K,
                'min_p': MIN_P,
                'seed': seed,
                'max_tokens': MAX_TOKENS,
                'messages': [{'role': 'system', 'content': SYSTEM}, {'role': 'user', 'content': prompt}],
            },
            ep.key,
            ep.model,
        )
    except HttpError as err:
        raise RuntimeError(f'{ep.model} {err.status}: {err.body[:400]}') from err
    message: Mapping[str, object] | None = dig(data, 'choices', 0, 'message')
    finish_reason: str | None = dig(data, 'choices', 0, 'finish_reason')
    answer, reasoning_text = separate_trace(message, finish_reason)
    if not answer.strip() and finish_reason != 'length':
        raise RuntimeError(f'{ep.model} returned no content (finish_reason {finish_reason})')
    return Reply(
        answer,
        reasoning_text,
        data.get('usage'),
        finish_reason,
        (time.perf_counter() - t0) * 1000,
        data.get('provider'),
        data.get('id'),
    )


def main() -> None:
    load_dotenv()
    endpoints = endpoints_from_env(os.environ)
    chunks = [Chunk(**c) for c in read_json_lines(CORPUS, required=True)]
    labels: list[Labelled] = read_json_lines(DATA / 'retrieval.labels.jsonl', required=True)
    selected = [label for label in labels if label['kind'] == SET]

    meta: EmbeddingsIndex = json.loads((DATA / 'embeddings.index.json').read_text(encoding='utf-8'))
    vectors = unpack_vectors((DATA / 'embeddings.bin').read_bytes(), meta['dim'], len(meta['ids']))
    qv: dict[str, list[float]] = json.loads((DATA / 'query-vectors.json').read_text(encoding='utf-8'))
    assert_dense_inputs_fresh(chunks, meta, [label['query'] for label in labels], qv)
    dense = DenseRetriever(meta['ids'], vectors, lambda q: qv[q])
    by_id = {c.id: c for c in chunks}

    done = done_keys(OUT)
    ran = missing_reasoning = truncated = 0
    for seed in SEEDS:
        for label in selected:
            context_ids = dense.search(label['query'], TOP_K)
            prompt = build_prompt(label['query'], [by_id[id] for id in context_ids])
            for ep in endpoints:
                if run_key(ep.model, label['query'], seed) in done:
                    continue
                reply = ask(prompt, ep, seed)
                reasoning_tokens = reasoning_tokens_of(reply.usage)
                if reasoning_tokens is None or (reasoning_tokens == 0 and reply.reasoning_text):
                    missing_reasoning += 1
                if reply.finish_reason == 'length':
                    truncated += 1
                row: Run = {
                    'query': label['query'],
                    'model': ep.model,
                    'seed': seed,
                    'answer': reply.answer,
                    'reasoningTokens': reasoning_tokens,
                    'reasoningChars': len(reply.reasoning_text),
                    'completionTokens': dig(reply.usage, 'completion_tokens'),
                    'promptTokens': dig(reply.usage, 'prompt_tokens'),
                    'latencyMs': reply.latency_ms,
                    'finishReason': reply.finish_reason,
                    'provider': reply.provider,
                    'generationId': reply.generation_id,
                    'contextIds': context_ids,
                }
                append_json_line(OUT, row)
                ran += 1
                cut = ' | USEKNUTO' if reply.finish_reason == 'length' else ''
                tokens = '—' if reasoning_tokens is None else reasoning_tokens
                chars = len(reply.reasoning_text)
                print(
                    f'{ran} [seed {seed}] {ep.model} | reasoning {tokens} tok / {chars} zn'
                    f' | completion {row["completionTokens"] or "—"}{cut}'
                )

    print(f'\nhotovo: {SET} set, seedy {",".join(map(str, SEEDS))}, {ran} běhů zapsáno do {OUT}')
    if missing_reasoning:
        print(f'POZOR: {missing_reasoning} odpovědí s nedůvěryhodným reasoning_tokens.')
        print('Použij reasoningChars; viz thinkingcap-preregistration.md, dodatek 3.')
    if truncated:
        print(f'POZOR: {truncated} odpovědí useknuto stropem {MAX_TOKENS} tokenů.')
        print('Ty otázky se ze srovnání vyřazují — viz dodatek 2, ne že by šetřily tokeny.')


if __name__ == '__main__':
    main()
